port/esp_idf/fw_update_esp_idf.c

```diff
@@ -104,8 +105,28 @@ enum golioth_status fw_update_post_download(void)
     return GOLIOTH_OK;
 }
 
-enum golioth_status fw_update_validate(void)
+enum golioth_status fw_update_validate(const uint8_t *hash, const size_t size)
 {
+    _update_partition = esp_ota_get_next_update_partition(NULL);
+    assert(_update_partition);
+    esp_image_metadata_t data;
+    const esp_partition_pos_t part_pos = {
+        .offset = _update_partition->address,
+        .size = _update_partition->size,
+    };
+    if (esp_image_verify(ESP_IMAGE_VERIFY, &part_pos, &data) != ESP_OK)
+    {
+        return GOLIOTH_ERR_FAIL;
+    }
+    if (data.image_len != size)
+    {
+        return GOLIOTH_ERR_FAIL;
+    }
+    if (memcmp(hash, data.image_digest, sizeof(data.image_digest)) != 0)
```

Comment:
Do we need to verify the length of `hash` is the same as the length of `data.image_digest`?

Reply:
We don't know the length of `hash` here -- it is somewhat implicit in the port API contract as implemented, so it may be worth making it more explicit that a 32 byte sha256 hash is being passed (at the very least in the documentation on `fw_update_validate`). In practice, because `data.image_digest` is [hard-coded to 32 bytes](https://github.com/espressif/esp-idf/blob/c5865270b50529cd32353f588d8a917d89f3dba4/components/bootloader_support/include/esp_image_format.h#L35) and we always pass a `GOLIOTH_OTA_COMPONENT_BIN_HASH_LEN` (32) then they should always be same length.

I am still in process of testing the esp-idf `fw_update_validate` implementation though, so there may need to be some further changes made here.

Reply:
Ah, thanks for the background.

>In practice, because data.image_digest is hard-coded to 32 bytes and we always pass a GOLIOTH_OTA_COMPONENT_BIN_HASH_LEN (32) then they should always be same length.

Would it make sense to change the function as below?

``` c
enum golioth_status fw_update_validate(const uint8_t hash[GOLIOTH_OTA_COMPONENT_BIN_HASH_LEN],
                                       const size_t size)
```

Reply:
I typically shy away from using arrays as function parameters in C because in practice they decay to a pointers. However, they can give the impression that, for example, `sizeof(hash)` in a function implementation would return the size of the array, but in reality it would return the size of the pointer. Some folks certainly argue the other side of this, so I'm open to discussion 🙂